compiler/z/codegen/UnaryEvaluator.cpp

```diff
@@ -310,21 +310,46 @@ OMR::Z::TreeEvaluator::dsqrtEvaluator(TR::Node * node, TR::CodeGenerator * cg)
    {
    TR::Node * firstChild = node->getFirstChild();
    TR::Register * targetRegister = NULL;
-   TR::Register * opRegister = cg->evaluate(firstChild);
 
-   if (cg->canClobberNodesRegister(firstChild))
+
+   // Calculate it for ourselves
+   if (firstChild->getOpCode().isLoadConst())
       {
-      targetRegister = opRegister;
+      union { double valD; int64_t valI; } result;
+      targetRegister = cg->allocateRegister(TR_FPR);
+      result.valD = sqrt(firstChild->getDouble());
+      TR::S390ConstantDataSnippet * cds = cg->findOrCreate8ByteConstant(node, result.valI);
+      generateRXInstruction(cg, TR::InstOpCode::LD, node, targetRegister, generateS390MemoryReference(cds, cg, 0, node));
```

Comment:
I honestly don't think we should be handling constant cases in these evaluators. IMO this should be the job of the optimizer (simplifier), not the codegen. This piece of code effectively duplicates what `dconstEvaluator` already does [1] which seems redundant. Not to mention if each codegen is doing this we are replicating the work several times.

@0xdaryl would love to hear your thoughts on this as it seems prevalent in many arithmetic evaluators.

My vote is to remove constant handling and if it doesn't exist in the optimizer implement it there so that all codegens get the benefit transparently.

[1] https://github.com/eclipse/omr/blob/cc820d3800101fb8dbf839921b26612ff9f07b6b/compiler/z/codegen/FPTreeEvaluator.cpp#L667-L676

Reply:
> IMO this should be the job of the optimizer (simplifier), not the codegen
I think @andrewcraik had mentioned something similar when I ran into him earlier.

I think the code for doing this is already available in the simplifier:
https://github.com/eclipse/omr/blob/703fd3bd7b24a568a3c1376a3db2137d7bd33500/compiler/optimizer/OMRSimplifierHandlers.cpp#L11659-L11683

We'll just need to make sure that it's working as expected.